pip/req/req_set.py

```diff
@@ -238,6 +238,15 @@ def add_requirement(self, install_req, parent_req_name=None):
                 existing_req = None
             if (parent_req_name is None and existing_req and not
                     existing_req.constraint):
+                if existing_req.name == install_req.name:
+                    # If the extras do not match, then union them together to
+                    # support install bar bar[foo]. If they do match, the
+                    # InstallationError will be raised as per normal.
+                    if existing_req.extras != install_req.extras:
+                        existing_req.extras = tuple(
+                            set(existing_req.extras).union(
+                                set(install_req.extras)))
+                        return [existing_req]
```

Comment:
It looks like if you `pip install foo[bar]==1.1 foo[baz]==1.2` you will end up, silently, with foo[bar,baz]==1.1 ?

This does not seem right to me, you should be comparing `canonicalize_name(install_req.req.specifier)` with `canonicalize_name(existing_req.req.specifier)`.